src/InvariantTheory/invariant_rings.jl

```diff
@@ -1,130 +1,166 @@
-mutable struct InvariantRing{T1 <: RingElem, T2 <: MatrixElem{T1}}
-   field::Nemo.Field # coefficient field
 
-   group::MatrixGroup{T1, T2}
-   matrices::Vector{T2} # Needs to be set (so far)
+mutable struct InvRing{S, T, U, V, W, X}
+   field::S
+   poly_ring::T
 
-   poly_ring::MPolyRing{T1}
+   group::U
+   action::Vector{V}          # Needs to be set (so far)
+   action_singular::Vector{W} # Needs to be set (so far)
 
-   primary::Vector{<: MPolyElem{T1}}
-   secondary::Vector{<: MPolyElem{T1}}
-   irreducible_secondary::Vector{<: MPolyElem{T1}}
+   modular::Bool
+
+   primary::Vector{X}
+   secondary::Vector{X}
+   irreducible_secondary::Vector{X}
+   fundamental::Vector{X}
 
    # Cache some stuff on the Singular side
    # (possibly removed at some point)
    reynolds_singular::Singular.smatrix
    molien_singular::Singular.smatrix
    primary_singular::Singular.smatrix
 
-   function InvariantRing{T1, T2}(matrices::Vector{T2}) where { T1 <: RingElem, T2 <: MatrixElem{T1} }
-      @assert !isempty(matrices)
-      IR = new{T1, T2}()
+   function InvRing(K::S, R::T, G::U, action::Vector{V}) where {S, T, U, V}
+     n = degree(G)
+     R, = PolynomialRing(K, "x" => 1:n)
+     R_sing = singular_ring(R)
+     action_singular = identity.([change_base_ring(R_sing, g) for g in action])
```

Comment:
One question @thofma: Why do you call `identity.` here? Actually, what exactly does it do?

Reply:
`identity.(v)` yields `v` with the tightest type possible:
```julia
julia> identity.(Any[1, 1.0])
2-element Vector{Real}:
 1
 1.0

julia> identity.(Any[1, 2])
2-element Vector{Int64}:
 1
 2
```
(I wanted to properly type the `[ ... ]` comprehension, but I don't know the type of the resulting matrices.)

Reply:
Cool! I didn't know about that.